### ISSUE TITLE
clingo concretizer not honoring version specs

### ISSUE THREAD

**Issue by glennpj** [LLNL/spack#27085]
### Steps to reproduce

There are quite a few open clingo related issues so this may have already been reported.

I can not build llvm-12 so I set llvm in `$spack/etc/packages.yaml`
```
packages:
...
  llvm:
    version: [11.1.0]
...
```
That is not honored when concretizing.
```console
$ spack spec mesa | grep llvm
mesa@21.2.3%gcc@9.4.0+glx+llvm+opengl~opengles+osmesa~strip buildtype=debugoptimized default_library=shared swr=auto arch=linux-centos7-x86_64
    ^llvm@12.0.1%gcc@9.4.0~all_targets+clang~code_signing+compiler-rt~cuda~flang+gold+internal_unwind~ipo+libcxx+lld+lldb~llvm_dylib~mlir+omp_as_runtime~omp_debug~omp_tsan+polly~python~shared_libs~split_dwarf build_type=Release cuda_arch=none arch=linux-centos7-x86_64
```
There is no indication as to why it is overriding my constraint. What it really should print is why it can not satisfy the constraint. If I switch the concretizer to "original" I can see what is wrong.
```console
$ spack spec mesa | grep llvm
==> Warning: the original concretizer is currently being used.
        Upgrade to "clingo" at your earliest convenience. The original concretizer will be removed from Spack starting at v0.18.0
==> Error: Conflicts in concretized spec "mesa@21.2.3%gcc@9.4.0+glx+llvm+opengl~opengles+osmesa~strip buildtype=debugoptimized default_library=shared swr=auto arch=linux-centos7-x86_64/wdggfum"

List of matching conflicts for spec:

    llvm@11.1.0%gcc@9.4.0~all_targets+clang~code_signing+compiler-rt~cuda~flang+gold+internal_unwind~ipo+libcxx+lld+lldb~llvm_dylib~mlir+omp_as_runtime~omp_debug~omp_tsan+polly~python~shared_libs~split_dwarf build_type=Release cuda_arch=none patches=6811ad441aa0221fb6f519550e7fc75a69a84998d44c1bea6d507605213c769c arch=linux-centos7-x86_64
        ^binutils@2.37%gcc@9.4.0~gas+gold~headers~interwork+ld~libiberty~lto+nls+plugins libs=shared,static arch=linux-centos7-x86_64
            ^diffutils@3.8%gcc@9.4.0 arch=linux-centos7-x86_64
            ^gettext@0.21%gcc@9.4.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-x86_64
                ^bzip2@1.0.8%gcc@9.4.0~debug~pic+shared arch=linux-centos7-x86_64
                ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-centos7-x86_64
                ^libxml2@2.9.12%gcc@9.4.0~python arch=linux-centos7-x86_64
                    ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-centos7-x86_64
                    ^xz@5.2.5%gcc@9.4.0~pic libs=shared,static arch=linux-centos7-x86_64
                    ^zlib@1.2.11%gcc@9.4.0+optimize+pic+shared arch=linux-centos7-x86_64
                ^ncurses@6.2.20210619%gcc@9.4.0+symlinks+termlib abi=6 arch=linux-centos7-x86_64
                ^tar@1.34%gcc@9.4.0 arch=linux-centos7-x86_64
        ^cmake@3.21.3%gcc@9.4.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-centos7-x86_64
        ^hwloc@2.5.0%gcc@9.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-centos7-x86_64
            ^libpciaccess@0.16%gcc@9.4.0 arch=linux-centos7-x86_64
                ^libtool@2.4.6%gcc@9.4.0 arch=linux-centos7-x86_64
                ^util-macros@1.19.3%gcc@9.4.0 arch=linux-centos7-x86_64
        ^libedit@3.1-20210216%gcc@9.4.0 arch=linux-centos7-x86_64
        ^perl-data-dumper@2.173%gcc@9.4.0 arch=linux-centos7-x86_64
            ^perl@5.34.0%gcc@9.4.0+cpanm+shared~threads arch=linux-centos7-x86_64
        ^python@3.9.7%gcc@9.4.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87,4c2457325f2b608b1b6a2c63087df8c26e07db3e3d493caf36a56f0ecf6fb768,f2fd060afc4b4618fe8104c4c5d771f36dc55b1db5a4623785a4ea707ec72fb4 arch=linux-centos7-x86_64
        ^swig@4.0.2%gcc@9.4.0 arch=linux-centos7-x86_64
            ^pcre@8.44%gcc@9.4.0~jit+multibyte+utf arch=linux-centos7-x86_64
        ^z3@4.8.9%gcc@9.4.0~gmp~ipo~python build_type=RelWithDebInfo arch=linux-centos7-x86_64

1. "+omp_as_runtime" conflicts with "llvm@:11.1" [omp_as_runtime works since LLVM 12.]
```
If I then set `~omp_as_runtime` as an llvm variant then the clingo spec is correct.
```console
$ spack spec mesa | grep llvm
mesa@21.2.3%gcc@9.4.0+glx+llvm+opengl~opengles+osmesa~strip buildtype=debugoptimized default_library=shared swr=auto arch=linux-centos7-x86_64
    ^llvm@11.1.0%gcc@9.4.0~all_targets+clang~code_signing+compiler-rt~cuda~flang+gold+internal_unwind~ipo+libcxx+lld+lldb~llvm_dylib~mlir~omp_as_runtime~omp_debug~omp_tsan+polly~python~shared_libs~split_dwarf build_type=Release cuda_arch=none patches=6811ad441aa0221fb6f519550e7fc75a69a84998d44c1bea6d507605213c769c arch=linux-centos7-x86_64
```
So, the constraint is not honored. The constraint should be honored and a message given as to why the spec fails.

### Error message

This looks like the relevant information with the `--debug` option.
```console
$ spack --debug spec mesa
==> [2021-10-30-04:15:26.960360] opt_criterion(16, deprecated versions used)
==> [2021-10-30-04:15:26.960475] opt_criterion(15, version weight)
==> [2021-10-30-04:15:26.960503] opt_criterion(14, number of non-default variants (roots))
==> [2021-10-30-04:15:26.960526] opt_criterion(13, preferred providers for roots)
==> [2021-10-30-04:15:26.960547] opt_criterion(12, number of values in multi valued variants (root))
==> [2021-10-30-04:15:26.960568] opt_criterion(11, number of non-default variants (non-roots))
==> [2021-10-30-04:15:26.960589] opt_criterion(9, preferred providers (non-roots))
==> [2021-10-30-04:15:26.960610] opt_criterion(8, compiler mismatches)
==> [2021-10-30-04:15:26.960630] opt_criterion(7, version badness)
==> [2021-10-30-04:15:26.960650] opt_criterion(6, number of values in multi valued variants (non-root))
==> [2021-10-30-04:15:26.960671] opt_criterion(5, non-preferred compilers)
==> [2021-10-30-04:15:26.960691] opt_criterion(4, target mismatches)
==> [2021-10-30-04:15:26.960711] opt_criterion(3, non-preferred targets)
==> [2021-10-30-04:15:26.970739] Ordered hashes [flex]: flex/151/09c22e5c6fef327d3e48eb23f0d610dcd3a35ab9207f12e0f875701c677978d3
==> [2021-10-30-04:15:26.971333] Ordered hashes [python]: python/27/0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87, python/28/f2fd060afc4b4618fe8104c4c5d771f36dc55b1db5a4623785a4ea707ec72fb4, python/30/4c2457325f2b608b1b6a2c63087df8c26e07db3e3d493caf36a56f0ecf6fb768
==> [2021-10-30-04:15:26.971451] Ordered hashes [meson]: meson/21/aa6c50d5a2aeb1a487d16f6712be4357fefb923aae37ab830699b07338388287
```

### Information on your system

* **Spack:** 0.16.3-5118-c38d3044fa
* **Python:** 3.9.7
* **Platform:** linux-centos7-cascadelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by zzzoom** [comment]
I've hit this bug today, or a very similar one. clingo isn't honoring any concretization preferences at all inside environments either:

```console
$ spack env activate -d .
$ cat spack.yaml
spack:
  config:
    concretizer: clingo
  specs: [amdscalapack@3.0]
  view: true
  packages:
    all:
      providers:
        blas: [amdblis]
        lapack: [amdlibflame]
$ spack concretize
==> Starting concretization
==> Environment concretized in 11.34 seconds.
==> Concretized amdscalapack@3.0
 -   p6htmyi  amdscalapack@3.0%gcc@11.2.0~ipo~pic+shared build_type=Release arch=linux-rocky8-zen2
[+]  vegcvgy      ^cmake@3.21.4%gcc@11.2.0~doc+ncurses~openssl+ownlibs~qt build_type=Release arch=linux-rocky8-zen2
[+]  pyruqqy          ^ncurses@6.1.20180224%gcc@11.2.0~symlinks+termlib abi=6 arch=linux-rocky8-zen2
 -   fke6jpg      ^openblas@0.3.18%gcc@11.2.0~bignuma~consistent_fpcsr~ilp64+locking+pic+shared threads=none arch=linux-rocky8-zen2
[+]  rbcc5nw          ^perl@5.26.3%gcc@11.2.0~cpanm+shared+threads patches=8cf4302ca8b480c60ccdcaa29ec53d9d50a71d4baf469ac8c6fca00ca31e58a2 arch=linux-rocky8-zen2
 -   vsnyrpd      ^openmpi@4.1.1%gcc@11.2.0~atomics~cuda~cxx~cxx_exceptions+gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker+pmi+pmix~singularity~sqlite3+static~thread_multiple+vt~wrapper-rpath fabrics=cma,hcoll,ucx,xpmem schedulers=slurm arch=linux-rocky8-zen2

==> Updating view at /home/bc/env/provider/.spack-env/view
$ perl -pi -e 's/clingo/original/mg' spack.yaml
$ spack concretize -f
==> Starting concretization
==> Warning: the original concretizer is currently being used.
        Upgrade to "clingo" at your earliest convenience. The original concretizer will be removed from Spack starting at v0.18.0
==> Environment concretized in 2.31 seconds.
==> Concretized amdscalapack@3.0
 -   ctcj67w  amdscalapack@3.0%gcc@11.2.0~ipo~pic+shared build_type=Release arch=linux-rocky8-zen2
 -   gvquwpq      ^amdblis@3.0%gcc@11.2.0+blas+cblas+shared+static threads=none arch=linux-rocky8-zen2
 -   7uoebyo          ^python@3.8.12%gcc@11.2.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87,4c2457325f2b608b1b6a2c63087df8c26e07db3e3d493caf36a56f0ecf6fb768,f2fd060afc4b4618fe8104c4c5d771f36dc55b1db5a4623785a4ea707ec72fb4 arch=linux-rocky8-zen2
[+]  binfkr4              ^bzip2@1.0.6%gcc@11.2.0~debug~pic+shared arch=linux-rocky8-zen2
[+]  bomaxnr              ^expat@2.4.1%gcc@11.2.0+libbsd arch=linux-rocky8-zen2
[+]  7hwwngk                  ^libbsd@0.11.3%gcc@11.2.0 arch=linux-rocky8-zen2
[+]  j564hui                      ^libmd@1.0.3%gcc@11.2.0 arch=linux-rocky8-zen2
[+]  za7qjkc              ^gdbm@1.21%gcc@11.2.0 arch=linux-rocky8-zen2
[+]  gyqba7j                  ^readline@8.1%gcc@11.2.0 arch=linux-rocky8-zen2
[+]  pyruqqy                      ^ncurses@6.1.20180224%gcc@11.2.0~symlinks+termlib abi=6 arch=linux-rocky8-zen2
[+]  n65wliy              ^gettext@0.21%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-rocky8-zen2
[+]  xmltze3                  ^libiconv@1.16%gcc@11.2.0 libs=shared,static arch=linux-rocky8-zen2
[+]  f7acl7c                  ^libxml2@2.9.12%gcc@11.2.0~python arch=linux-rocky8-zen2
[+]  uvqcawm                      ^pkgconf@1.8.0%gcc@11.2.0 arch=linux-rocky8-zen2
[+]  li2m5sr                      ^xz@5.2.5%gcc@11.2.0~pic libs=shared,static arch=linux-rocky8-zen2
[+]  ov2nlgl                      ^zlib@1.2.11%gcc@11.2.0+optimize+pic+shared arch=linux-rocky8-zen2
[+]  x3dcaby                  ^tar@1.34%gcc@11.2.0 arch=linux-rocky8-zen2
[+]  klgmr7w              ^libffi@3.3%gcc@11.2.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-rocky8-zen2
 -   j3k4nch              ^openssl@1.1.1g%gcc@11.2.0~docs certs=system arch=linux-rocky8-zen2
[+]  2rp7vw6              ^sqlite@3.36.0%gcc@11.2.0+column_metadata+fts~functions~rtree arch=linux-rocky8-zen2
[+]  c2pvezt              ^util-linux-uuid@2.36.2%gcc@11.2.0 arch=linux-rocky8-zen2
 -   4ddtkf7      ^amdlibflame@3.0%gcc@11.2.0~debug+lapack2flame+shared+static patches=b3066e8ea70f9a59d1ce00330d72764482dd0faa57d185a45f73ce0effa2bc14 threads=none arch=linux-rocky8-zen2
 -   ulafzdi      ^cmake@3.21.4%gcc@11.2.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-rocky8-zen2
 -   53jopvb      ^openmpi@4.1.1%gcc@11.2.0~atomics~cuda~cxx~cxx_exceptions+gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker+pmi+pmix~singularity~sqlite3+static~thread_multiple+vt~wrapper-rpath fabrics=ucx schedulers=slurm arch=linux-rocky8-zen2

==> Updating view at /home/bc/env/provider/.spack-env/view
==> Warning: [/home/bc/env/provider/.spack-env/._view/vjynr5er5mrnowqnypbl7fkxjo2o2pqt] Skipping external package: ncurses@6.1.20180224%gcc@11.2.0~symlinks+termlib abi=6 arch=linux-rocky8-zen2/pyruqqy
==> Warning: [/home/bc/env/provider/.spack-env/._view/vjynr5er5mrnowqnypbl7fkxjo2o2pqt] Skipping external package: bzip2@1.0.6%gcc@11.2.0~debug~pic+shared arch=linux-rocky8-zen2/binfkr4
```

**Comment by haampie** [comment]
@zzzoom can't reproduce, what commit were you on?

**Comment by alalazo** [comment]
> So, the constraint is not honored. The constraint should be honored and a message given as to why the spec fails.

I'm not sure I would classify this as a bug. Version, variants, etc. in `packages.yaml` are not constraints - they are just preferences. Spack is told to prefer version 11.1.0 if there are no "issues" with it, but is free to choose anything else if that version cannot be used or leads to a "worse" solution (in terms of the current optimization targets that clingo has).

For dependencies, minimizing the number of non-default variants is a higher priority target for clingo than minimizing version badness - hence the selection of version 12.0.1 If you add a variant preference:
```yaml
packages:
  llvm:
    version: ["11.1.0"]
    variants: '~omp_as_runtime'
```
then Spack will choose `llvm@11.1.0` since `~omp_as_runtime` becomes the default choice and is not penalized anymore in the optimization.

**Comment by zzzoom** [comment]
> @zzzoom can't reproduce, what commit were you on?

f45ef21e375d2a57813d64818d055b89582ebdc5

I'll try to reproduce it in a cleaner install.

**Comment by glennpj** [comment]
> I'm not sure I would classify this as a bug. Version, variants, etc. in `packages.yaml` are not constraints - they are just preferences.

Okay, constraint is not the correct term, sorry. I do consider this a bug though as per the issue template:
```
Report a bug in the core of Spack (command not working as expected, etc.)
```

Essentially, the original concretizer treated versions in `packages.yaml` more like constraints. However, the behavior with clingo is very different from that of the original concretizer and I prefer the behavior of the original concretizer here. The original concretizer tries to use my preference and informs me of a conflict. From that I inferred that was the reason why clingo went to llvm-12. Without that behavior, and subsequent information from the original concretizer, getting the correct specs would have been much more difficult.

> Spack will choose llvm@11.1.0 since ~omp_as_runtime becomes the default choice and is not penalized anymore in the optimization.

Yes, but I did not know that was what was causing `clingo` to bump to llvm-12 over my "preference" of llvm-11 until after I switched back to the original concretizer. If the behavior is different but correct going forward then we can change this issue from a "bug" to a "feature request" to stop and post a message when a version preference in `packages.yaml` can not be met. Perhaps there is a way to do that and I am simply unaware.
